### PR TITLE
Editor Welcome Tour Component

### LINKED ISSUE #47691: MVP Welcome Tour Component
## Objectives
MVP Welcome Tour Component

### As a user I 
- [ ] see Tour Component when opening editor for first timeg (in a future issue: Tour also accessible via MoreMenu)
- [ ] do not see the tour component when opening editor subsequent times (save state)
- [ ] can minimize & maximize Tour UI
- [ ] can interact with both the Tour and the Editor
- [ ] may see conditional content (foundations should allow for this in the future e.g. tailored onboarding experience for types of users)
- [ ] progress is tracked similar to [NUX model](https://github.com/Automattic/wp-calypso/blob/6a9bbcc26900d07952e93de8255433226533ffd3/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-nux%2Fsrc%2Fwpcom-nux.js#L20) tracking

### Tech ToDos

- [ ] Preload images to avoid janky loading of card content

## Resources
pbAok1-1Dh-p2
Design: p9Jlb4-1Vj-p2

### PR DESCRIPTION
**Pull request by autumnfjeld**
This PR continues from [initial spike](https://github.com/Automattic/wp-calypso/issues/47691).
[Project issue](https://github.com/Automattic/wp-calypso/issues/47691) details user facing requirements and project links.

## Changes proposed in this Pull Request
* Add new Welcome Tour Component
* Remove NUX modal

#### Details:

- [x] create portal for the Tour component such that user can interact with either Tour or the Editor
- [x] Tour card can be minimized and then re-maximized
- [x] pagination for Tour cards
- [x] foundations that allow for conditional content (e.g. welcome content customized for different onboarding experiences)
- [x] user can interact in both Tour & editor 
- [x] Set up correct icons / discuss with design as icons should come from [GB icons package ](https://github.com/WordPress/gutenberg/blob/ad97916b957307ec592b53d7d6b8874211cca167/packages%2Ficons%2Fsrc%2Flibrary%2Falign-center.js#L12-L13) .
- [x] images should be preloaded to avoid janky renders
- [x] wire up wp.data things to save state: isNuxEnabled to know when to show the Tour
- [x] verify that we can repurpose [ `is_nux_enabled`](https://github.com/Automattic/wp-calypso/blob/720b5dad4f5217adc9c1de3e2d706b56f6bb7cf9/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-nux%2Fclass-wp-rest-wpcom-block-editor-nux-status-controller.php#L76) or if we need to create another flag a different flag for tracking when editor is opened for the first time
- [x] ~remove old NUX modal code UNLESS we are doing an A/B test~
- [x] Move Tour to own folder to allow for A/B testing of NUX vs Tour
- [ ] translations: TODO in followup PR
- [ ] tracking: TODO in followup PR

#### Issues:
* tour not loading in safari
* Not clear if the `<LaunchWpcomNuxTour />` component is unmounting. I did not observe the useEffect cleanup happening.

## Testing instructions
There are three places used to determine whether to show the NUX dialog, in the following order:
1) Local storage in the browser
2) The user attribute `wpcom-block-editor-preferences['nux-status']`
3) The user metadata field `wpcom_block_editor_nux_status`

### 1) reset local storage

You must  clear local storage for the site you are working with (from the Application tab of the chrome dev tools)
Screen Shot 2020-12-04 at 5 01 13 PM

### 2) Reset the user attribute `wpcom-block-editor-preferences['nux-status']`
The user attribute 'wpcom-block-editor-preferences' => 'nux-status' is used to determine whether to show the welcome tour.
This flag is also cached in local storage in the browser
It can be reset like so from a sandbox: (for your user_id)
```
wpsh
> update_user_attribute( 180278607, 'wpcom-block-editor-preferences', [ 'nux-status' => 'enabled' ] );
```
At this point reloading the page should show the tour again

### 3) Reset the user metadata field `wpcom_block_editor_nux_status`
This field is only used if the user attribute hasn't been set yet
```
wp user meta update <USER_NAME> wpcom_block_editor_nux_status enabled
```

Fixes https://github.com/Automattic/wp-calypso/issues/47691

### CODE REVIEW

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/tour-content.js
Comment (a8ci18n) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/tour-content.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **130** times:
`translate( 'Welcome to WordPress!' )` *ES Score: 8*

**Comment by dubielzyk** [comment]
Looking super good. Here's a few pixel changes:
- [x] Let's make the `.wpcom-editor-welcome-tour-frame` position `left: 16px` instead of `40px`. And `bottom: 44px` instead of `40px`
- [x] The close and minimize button should only appear once the Welcome Tour is hovered. The buttons should have opacity 0.7 and 0.9 on the button hover. [See gif](https://d.pr/i/j5FWcX)
- [x] The close and minimize button are currently `36px`, but should be `32px`. The icon within should be 24px.
- [x] Remove the current border and box-shadow on the Welcome Tour and use `box-shadow: rgba(0, 0, 0, 0.1) 0px 0px 0px 1px, rgba(0, 0, 0, 0.1) 0px 2px 4px 0px` instead
- [x] The `components-card__body` has a padding of `12px` but it should be `16px`
- [x] The [active step color](https://d.pr/i/gGdCjG) shouldn't be blue, but rather `gray 700` (or equivalent)
- [x] The copy says `Welcome to Wordpress`, but it should be `Welcome to WordPress`
- [x] `welcome-tour-card__heading` should have `font-size: 1.125rem` not `1.25rem`
- [x] The distance between the heading and the text is ~24px, but should be ~16px. [Mock](https://d.pr/i/6MYSG0)

**Comment by roo2** [comment]
Thanks for the feedback @dubielzyk I was able to address them all, let's make a build of it on monday and upload it to the ephemeral site for testing.

I tried to test the welcome tour in safari but had a few problems:
* I can't seem to get my sandbox to work in safari, my hostsfile generally does work but when I set public-api to point to my sandbox, for some reason I still see the production api in safari... so I wasn't able to test this.
* I tried uploading a build of the plugin to an atomic site but it seems that atomic sites use different logic to determine whether to show the welcome tour, perhaps it doesn't even make sense to show the welcome tour to atomic users since presumably they would be more advanced users already?

I applied a bit of a simple fix for the janky image loading, I just put a fixed height on the image container so that the welcome tour doesn't jump around as images are loading, you might still look into preloading images if you think that's necessary as well.

I had intended to put this behind a feature toggle but we don't actually have access to our calypso feature flags from the editing toolkit plugin, we should find a way to A/B test this, we'll have to look into hat actual A/B testing options we have access to from the ETK plugin...

**Comment by ramonjd** [comment]
> we'll have to look into hat actual A/B testing options we have access to from the ETK plugin...

Maybe checkout what @deBhal and @andrewserong are doing on the premium patterns side of things. See pbmo2S-Ca-p2#comment-1369 and D53680-code

**Comment by dubielzyk** [comment]
This is great progress. Two more things:

- [ ] Let's remove the `1px solid rgb(226, 228, 231)` border on the `.components-card` div.
- [ ] This might cause the [cut corner look](https://d.pr/i/Py3ama). If that happens lets round the corners on the corners that are affected

**Comment by ramonjd** [comment]
>  I can't seem to get my sandbox to work in safari, my hostsfile generally does work but when I set public-api to point to my sandbox, for some reason I still see the production api in safari... so I wasn't able to test this.

I think I figured out why. See pbAok1-1Dh-p2#comment-3502

**Comment by ramonjd** [comment]
Let's get this in the ETK, even if we're not shipping. 

One way to do that would be to create a new directory for this work and copy the files over so we're not modifying the current NUX tour.

Then update the [load_wpcom_block_editor_nux()](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L253) method to conditionally load the new tour. 

For example:

```
/**
 * Load WPCOM Block Editor NUX
 */
function load_wpcom_block_editor_nux() {
	if (
		( defined( 'WPCOM_NUX_WELCOME_TOUR' ) && WPCOM_NUX_WELCOME_TOUR ) ||
		apply_filters( 'a8c_enable_wpcom_welcome_tour', false )
	) {
                // THE NEW ONE! 
		require_once __DIR__ . '/wpcom-block-editor-welcome-tour/class-wpcom-block-editor-welcome-tour.php';
                return;
	}
         // THE CURRENT ONE
	require_once __DIR__ . '/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php';
}
add_action( 'plugins_loaded', __NAMESPACE__ . '\load_wpcom_block_editor_nux' );
```

So, if `WPCOM_NUX_WELCOME_TOUR` constant is defined in the sandbox we'll load the new plugin. Later, we'd rely on the `a8c_enable_wpcom_welcome_tour` filter to control access on production. See D53894-code for an example

Some discussion here:
p1607396394436800-slack-CRWCHQGUB

Let me know if you have any questions!

**Comment by autumnfjeld** [comment]
@dubielzyk 
> * [ ]  Let's remove the `1px solid rgb(226, 228, 231)` border on the `.components-card` div.

Done.  Corners [look good](https://d.pr/i/bpB90M)

File: apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php
Comment (autumnfjeld) [inline apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php:246]: oops.  missing return.

File: apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php
Comment (ramonjd) [inline apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php]: Before we merge we can replace this `true` with something like the following:

```suggestion
	if ( defined( 'WPCOM_WELCOME_TOUR' ) && WPCOM_WELCOME_TOUR ||
		apply_filters( 'a8c_enable_wpcom_welcome_tour', false )
		) {
```
We did this [elsewhere](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L335).

The reason is that we can merged the code and, so long as we haven't turned on the `a8c_enable_wpcom_welcome_tour` filter in WPCOM, we can continue to hide the plugin behind the `WPCOM_WELCOME_TOUR` constant, which we can define on our sandboxes for testing.

Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php]: Yes!! Was already setting up the `WPCOM_WELCOME_TOUR` conditional. :)

I also wanted to set that constant on my ephemeral site, but am unable to ssh into it. So I just need to make a build with this `if (true)` hack for the ephemeral site, then I'll remove the hacks for merging.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php]: This is throwing a duplicate class name warning in PHPCS, because it shares the same name as the class defined in `wpcom-block-editor-nux` (https://github.com/automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php)

Is there a clean way we could share the one file and class between these two areas, to remove the duplication? It looks like they're very similar.

From PHPCS:

```
----------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------
 13 | WARNING | Duplicate class name
    |         | "A8C\FSE\WP_REST_WPCOM_Block_Editor_NUX_Status_Controller"
    |         | found; first defined in
    |         | /home/runner/work/wp-calypso/wp-calypso/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php
    |         | on line 13
    |         | (Generic.Classes.DuplicateClassName.Found)
----------------------------------------------------------------------
```

**Review by andrewserong — COMMENTED** [review]
This feature is looking so good @autumnfjeld!

I've left a comment about a duplicate class name warning in PHPCS because of the duplication of the NUX status controller. I can think of a couple of potential ways of dealing with this:

* If you don't mind the duplication with this class, you could rename it. Or, find a place to keep the shared PHP files between this feature and `wpcom-block-editor-nux` so it only needs to be defined once (this would be better so that there aren't two attempts to register effectively the same API endpoint). Or,

* Move the Editor Welcome Tour feature to be part of the `wpcom-block-editor` directory, and share code, which means you could also remove the duplication of the `disable-core-nux.js` and `public-path.js` files. The downside of this is that it mightn't be as neat to switch the feature and on off. Because we're enqueueing the one JS bundle, to handle switching features on and off we might need to use a `wp_localize_script` call to pass a boolean to the JS to flag that the feature is enabled. E.g.

```php
wp_localize_script(
  'wpcom-block-editor-nux-script',
  'wpcomBlockEditorWelcomeTourActivated',
  true, // the boolean value of whether or not the feature is enabled
);
```

Then, in the JS, you could detect whether or not the feature is enabled by checking `window.wpcomBlockEditorWelcomeTourActivated`.

Personally I don't mind a bit of duplication at all while we're exploring a new feature, it can make it much easier to try things without having to worry too much about affecting an existing feature, but it can also be a bit tricky when there are collisions (like with the API endpoint). So I think it'd be perfectly valid to keep the two separate if you think it makes sense for the experiment!

**Comment by autumnfjeld** [comment]
@andrewserong Thanks for all the suggestions!! :) 

I think I'm ok with leaving the duplicate javascript files (public-path.js, store.js, etc) because as you said, the state of the NUX and Welcome Tour code is currently exploration. 

I fixed the PHP duplication however.

> Move the Editor Welcome Tour feature to be part of the wpcom-block-editor directory

I assume you mean `wpcom-block-editor-nux directory`.  This is how I started the work, but we needed to separate the Tour from the NUX to run A/B tests. I think because php is in charge of determining [which component to show](https://github.com/Automattic/wp-calypso/blob/87dfbb241b6fefdfb57dff08b854dc3f2f765c64/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L245): `apply_filters( 'a8c_enable_wpcom_welcome_tour', false ) ` .  However, now that I think more deeply about this it could have probably been set up differently.....but maybe other ways would have been more complicated.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php
Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php]: I removed the duplicate file, keeping apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php (the one in the new wpcom-block-editor-welcome-tour folder)  And then updated the file path where the file is called in class-wpcom-block-editor-nux.php.

Does that seem like a sensible solution?

**Comment by andrewserong** [comment]
That sounds like a good trade-off to me @autumnfjeld, and nice idea getting `wpcom-block-editor-nux` to use the file in your PR so you've got a bit more flexibility for this feature 👍

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php:80]: I forgot to add: because the other usage of this controller also attempts to register the route, we might just want to test that this doesn’t throw an error anywhere. I think you can safely call to register the same route twice, but we might want to double check (check that it doesn’t throw errors and that the endpoint behaves as we’d expect). The function register_rest_route says that it returns false if it can’t register an endpoint so I suspect it’ll safely silently fail the second time, or just override the first call, which is also fine 🤞

Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php:80]: @andrewserong Thanks so much for pointing this out!  Got me to think/learn a bit more in depth about some WP & php concepts.

> register the same route twice

If I understand you correctly, the concern is that `
class WP_REST_WPCOM_Block_Editor_NUX_Status_Controller` will be registered twice, once when NUX code loads and again when Tour code loads.   However, only one of these will load at a time, it is an either/or situation for A/B testing.  This [code in full-site-editing-plugin.php](https://github.com/Automattic/wp-calypso/blob/137073582964d86579a57c9f85b63ed70b0af831/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L243) decides which to load.

In my sandbox I tested both scenarios and I didn't observe any errors in my `/tmp/php-errors` file or in dev tools. (Note: Adding `define( 'SHOW_WELCOME_TOUR', true );` to 0-sandbox.php file will force the new Tour to show). 

Does that address the issue? Or is there anything I'm missing/not understanding?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/style-tour.scss
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/style-tour.scss]: Is this still a TODO? I wasn't sure what might need to be changed in a follow-up?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js]: Can this comment be removed?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js]: Does `PaginationControl` require an empty className?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js
Comment (andrewserong) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:43]: Shall we add in the `__()` call to translate these strings of text now, or did you want to hold off until later, if the words are still being decided on?

**Review by andrewserong — APPROVED** [review]
This is looking great and testing so well @autumnfjeld!

I've left a couple of additional comments, but no blockers, and feel free to do cleanup work in follow-up PRs if it's easier so you can get the bulk of the work in, particularly as this is safely guarded behind the constant / filter check.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php
Comment (andrewserong) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wpcom-block-editor-nux.php:80]: @autumnfjeld perfect, that sounds good to me! And from reading [the docs](https://developer.wordpress.org/reference/functions/register_rest_route/) for `register_rest_route`, it looks like even if it wasn't guarded behind that either/or condition we'd probably still be safe 👍

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js
Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:43]: We'll do `__( )` calls in a subsequent PR.

Comment (andrewserong) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:43]: Sounds good!

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/style-tour.scss
Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/style-tour.scss]: removed. :)

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js
Comment (autumnfjeld) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js]: removed.

**Pull request merged**
